src/H5FDonion_index.c

```diff
@@ -0,0 +1,935 @@
+ * Function:    H5FD__onion_archival_index_is_valid
+ *
+ * Purpose:     Determine whether an archival index structure is valid.
+ *
+ *              + Verify page size (power of two).
+ *              + Verify list exists.
+ *              + Verify list contents:
+ *                + Sorted by increasing logical address (no duplicates)
+ *                + Logical addresses are multiples of page size.
+ *
+ * Return:      TRUE/FALSE
+ *-----------------------------------------------------------------------------
+ */
+hbool_t
+H5FD__onion_archival_index_is_valid(const H5FD_onion_archival_index_t *aix)
+{
+    hbool_t ret_value = TRUE;
+
+    FUNC_ENTER_PACKAGE_NOERR;
+
+    HDassert(aix);
+
+    if (H5FD__ONION_ARCHIVAL_INDEX_VERSION_CURR != aix->version)
+        HGOTO_DONE(FALSE)
+    if (NULL == aix->list)
+        HGOTO_DONE(FALSE)
+
+    /* Ensure list is sorted on logical_page field */
+    if (aix->n_entries > 1)
+        for (uint64_t i = 1; i < aix->n_entries - 1; i++)
+            if (aix->list[i + 1].logical_page <= aix->list[i].logical_page)
+                HGOTO_DONE(FALSE)
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value);
+} /* end H5FD__onion_archival_index_is_valid() */
+
+/*-----------------------------------------------------------------------------
+ * Function:    H5FD__onion_archival_index_find
+ *
+ * Purpose:     Retrieve the archival index entry by logical page ID.
+ *
+ *              The archival index pointer must point to a valid index entry.
+ *              The entry out pointer-pointer cannot be null.
+ *
+ * Return:      Success: Positive value (1) -- entry found.
+ *                       Entry out pointer-pointer is set to point to entry.
+ *              Failure: Zero (0) -- entry not found.
+ *                       Entry out pointer-pointer is unmodified.
+ *-----------------------------------------------------------------------------
+ */
+int
+H5FD__onion_archival_index_find(const H5FD_onion_archival_index_t *aix, uint64_t logical_page,
+                                const H5FD_onion_index_entry_t **entry_out)
+{
+    uint64_t                  low       = 0;
+    uint64_t                  high      = 0;
+    uint64_t                  n         = 0;
+    uint64_t                  range     = 0;
+    H5FD_onion_index_entry_t *x         = NULL;
+    int                       ret_value = 0;
+
+    FUNC_ENTER_PACKAGE_NOERR;
+
+    HDassert(aix);
+    HDassert(H5FD__ONION_ARCHIVAL_INDEX_VERSION_CURR == aix->version);
+    HDassert(entry_out);
+    if (aix->n_entries != 0)
+        HDassert(aix->list);
+
+    high  = aix->n_entries - 1;
+    range = high;
+
+    /* Trivial cases */
+    if (aix->n_entries == 0 || logical_page > aix->list[high].logical_page ||
+        logical_page < aix->list[0].logical_page)
+        HGOTO_DONE(0)
+
+    /*
+     * Binary search on sorted list
+     */
+
+    /* Winnow down to first of found or one element */
+    while (range > 0) {
```

Comment:
Seems we could replace some or all of these binary searches with bsearch, but that's just a nitpick